### DISCUSSION TITLE
Windows - unable to load dll - they are all there though

### DISCUSSION THREAD

**Discussion by andreas-it-dev** [libvips/ruby-vips#364] [Q&A]
Hi, 
i've got an issue on windows 11 and am unsure if i screwed up, if it is a bug or some weird windows thing.. 

this is the error message when trying to process an image from within rails:

```
VIPS-WARNING **: 06:59:07.223: unable to load "C:\Ruby31-x64\msys64\ucrt64/lib/vips-modules-8.13\vips-heif.dll" -- 'C:\Ruby31-x64\msys64\ucrt64/lib/vips-modules-8.13\vips-heif.dll': Das angegebene Modul wurde nicht gefunden.

GModule-CRITICAL **: 06:59:07.224: g_module_make_resident: assertion 'module != NULL' failed

VIPS-WARNING **: 06:59:07.236: unable to load "C:\Ruby31-x64\msys64\ucrt64/lib/vips-modules-8.13\vips-jxl.dll" -- 'C:\Ruby31-x64\msys64\ucrt64/lib/vips-modules-8.13\vips-jxl.dll': Das angegebene Modul wurde nicht gefunden.

GModule-CRITICAL **: 06:59:07.236: g_module_make_resident: assertion 'module != NULL' failed

VIPS-WARNING **: 06:59:07.248: unable to load "C:\Ruby31-x64\msys64\ucrt64/lib/vips-modules-8.13\vips-magick.dll" -- 'C:\Ruby31-x64\msys64\ucrt64/lib/vips-modules-8.13\vips-magick.dll': Das angegebene Modul wurde nicht gefunden.

GModule-CRITICAL **: 06:59:07.248: g_module_make_resident: assertion 'module != NULL' failed

VIPS-WARNING **: 06:59:07.260: unable to load "C:\Ruby31-x64\msys64\ucrt64/lib/vips-modules-8.13\vips-openslide.dll" -- 'C:\Ruby31-x64\msys64\ucrt64/lib/vips-modules-8.13\vips-openslide.dll': Das angegebene Modul wurde nicht gefunden.

GModule-CRITICAL **: 06:59:07.260: g_module_make_resident: assertion 'module != NULL' failed

VIPS-WARNING **: 06:59:07.273: unable to load "C:\Ruby31-x64\msys64\ucrt64/lib/vips-modules-8.13\vips-poppler.dll" -- 'C:\Ruby31-x64\msys64\ucrt64/lib/vips-modules-8.13\vips-poppler.dll': Das angegebene Modul wurde nicht gefunden.

GModule-CRITICAL **: 06:59:07.273: g_module_make_resident: assertion 'module != NULL' failed
Completed 500 Internal Server Error in 2051ms (ActiveRecord: 13.3ms | Allocations: 295138)
```

however, the files are all there:

![image](https://user-images.githubusercontent.com/13486613/226151461-a16c0d21-4365-4263-a75b-e845223cdaa9.png)

if that is important: vips sits at `c:\vips-dev`:

![image](https://user-images.githubusercontent.com/13486613/226151526-58fd1d53-b959-4736-ab1c-533d4358b4b6.png)

and i have `c:\vips-dev\bin` in my path:

![image](https://user-images.githubusercontent.com/13486613/226151607-87d62509-216d-4030-b9e5-168f02578b51.png)

any help is highly appreciated.

thanks

**Comment by andreas-it-dev** [comment]
did some more tests. vips itself works, but fails via ruby

![image](https://user-images.githubusercontent.com/13486613/226151957-a89547c8-20e6-4dc3-8c69-1aa1cea8f1c0.png)

**Comment by kleisauke** [comment]
I was able to reproduce this, it looks like a packaging issue in MSYS2 where the optional dynamic modules are distributed when you install libvips via `pacman` with:
```console
$ pacman -S mingw-w64-ucrt-x86_64-libvips
```
(which is done automatically during `gem install ruby-vips`, see #332)

But the dependencies of these optional dynamic modules are not installed. To fix this you can do either:
```console
$ pacman -S --asdeps mingw-w64-ucrt-x86_64-libheif mingw-w64-ucrt-x86_64-libjxl mingw-w64-ucrt-x86_64-imagemagick mingw-w64-ucrt-x86_64-openslide mingw-w64-ucrt-x86_64-poppler
```

Or simply remove (or rename) the directory where these dynamic modules resides (i.e. `C:\Ruby31-x64\msys64\ucrt64\lib\vips-modules-8.13`).

Note that if you want to use the libvips from https://github.com/libvips/build-win64-mxe, you can do:
```console
$ pacman -Rcns mingw-w64-ucrt-x86_64-libvips
```

And set the `RUBY_DLL_PATH` env accordingly (e.g. to `C:\vips-dev-8.14\bin`).

**Reply by jcupitt** [reply]
Ah that makes more sense. That PR is probably still useful -- we shouldn't be calling `make_resident` if module load fails.
